test/parallel/test-worker-eval-typescript.js

```diff
@@ -0,0 +1,76 @@
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const { Worker } = require('worker_threads');
+const { test } = require('node:test');
+
+const esmHelloWorld = `
+    import worker from 'worker_threads';
+    const foo: string = 'Hello, World!';
+    worker.parentPort.postMessage(foo);
+`;
+
+const cjsHelloWorld = `
+    const { parentPort } = require('worker_threads');
+    const foo: string = 'Hello, World!';
+    parentPort.postMessage(foo);
+`;
+
+const disableTypeScriptWarningFlag = '--disable-warning=ExperimentalWarning';
+
+test('Worker eval module typescript without input-type', () => {
+  const w = new Worker(esmHelloWorld, { eval: true, execArgv: [disableTypeScriptWarningFlag] });
+  w.on('message', common.mustCall((message) => {
+    assert.strictEqual(message, 'Hello, World!');
+  }));
+});
```

Comment:
This pattern for these tests is not going to work. If the assertion fails, the test results will include a warning but the test itself will not fail... for instance, if I change the `Hello, World!' in line 24 here to `Hello, World!!`, the test output becomes:

```
✔ Worker eval module typescript without input-type (1.629961ms)
✔ Worker eval module typescript with --input-type=module-typescript (0.501763ms)
✔ Worker eval module typescript with --input-type=commonjs-typescript (0.427135ms)
✔ Worker eval module typescript with --input-type=module (1.430576ms)
✔ Worker eval commonjs typescript without input-type (0.417876ms)
✔ Worker eval commonjs typescript with --input-type=commonjs-typescript (0.33923ms)
✔ Worker eval commonjs typescript with --input-type=module-typescript (0.481868ms)
ℹ Error: Test "Worker eval module typescript without input-type" at test/parallel/test-worker-eval-typescript.js:21:1 generated asynchronous activity after the test ended. This activity created the error "AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
actual expected

'Hello, World!!'
" and would have caused the test to fail, but instead triggered an uncaughtException event.
ℹ tests 7
ℹ suites 0
ℹ pass 7
ℹ fail 0
ℹ cancelled 0
ℹ skipped 0
ℹ todo 0
ℹ duration_ms 64.075491
```

In order for the test to fail correctly, using the non-promise form of the API, you'll need to use the `done` callback correctly inside the `message` event handler... something like...

```js
test('Worker eval module typescript without input-type', (t, done) => {
  const w = new Worker(esmHelloWorld, { eval: true, execArgv: [disableTypeScriptWarningFlag] });
  w.on('message', common.mustCall((message) => {
    try {
      assert.strictEqual(message, 'Hello, World!');
      done();
    } catch (err) {
      done(err);
    }
  }));
});
```

Which would cause the assertion to be properly reported as a test failure:

```
✖ Worker eval module typescript without input-type (52.120739ms)
  AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
  actual expected
  
  'Hello, World!!'
  
      at Worker.<anonymous> (/home/jsnell/projects/node/node/test/parallel/test-worker-eval-typescript.js:25:14)
      at Worker.<anonymous> (/home/jsnell/projects/node/node/test/common/index.js:491:15)
      at Worker.emit (node:events:507:28)
      at MessagePort.<anonymous> (node:internal/worker:267:53)
      at [nodejs.internal.kHybridDispatch] (node:internal/event_target:827:20)
      at MessagePort.<anonymous> (node:internal/per_context/messageport:23:28) {
    generatedMessage: true,
    code: 'ERR_ASSERTION',
    actual: 'Hello, World!',
    expected: 'Hello, World!!',
    operator: 'strictEqual'
  }

✔ Worker eval module typescript with --input-type=module-typescript (0.497239ms)
✔ Worker eval module typescript with --input-type=commonjs-typescript (0.403381ms)
✔ Worker eval module typescript with --input-type=module (0.391807ms)
✔ Worker eval commonjs typescript without input-type (0.413241ms)
✔ Worker eval commonjs typescript with --input-type=commonjs-typescript (0.40354ms)
✔ Worker eval commonjs typescript with --input-type=module-typescript (0.425815ms)
ℹ tests 7
ℹ suites 0
ℹ pass 6
ℹ fail 1
ℹ cancelled 0
ℹ skipped 0
ℹ todo 0
ℹ duration_ms 101.019681
```

This holds true for all of the test cases in this file.

Reply:
refactored with promise api